### ISSUE TITLE
Automatically notify support and discussion channels on release

### ISSUE THREAD

**Issue by BlueDrink9** [autokey/autokey#641]
Update the CI pipeline to message the gitter chat on releases, and to message the support list on release of a stable version.

Would go at the end of the build->release pipeline. Will probably need bots for those groups?

**Comment by josephj11** [comment]
@BlueDrink9 Didn't see how to contact you more directly. We have a potential new developer for AutoKey. Would you be willing to talk with him to help him get started?  He just showed up a day or two ago on Gitter, so we don't know much about him yet, but he doesn't seem to be trying to get a bounty, etc.

I'd also like to hear about what you are up to now. Hope all is well. I posted a few messages to you on Gitter with no replies.

**Comment by danballweg** [comment]
Looking at the Matrix API docs this looks like it is pretty straight forward.  Do you already have a token for the gitter app saved in the git secrets for the repo? I just need the secret name if you do, otherwise happy to chat anytime to help get one added if needed.
Also, is there anything specific that you want the message to say in gitter?  I was thinking something like:
"Release X.X.X built successfully"
or something like that. We could dump whatever you want in the message

**Comment by josephj11** [comment]
Whoever setup our Gitter page is long gone. I have never heard of that token. So we'd have create it. We would also want the message in our Google  groups forum . That should not be too difficult once there's an account it can use to post there.

We'll have to see what others say, but that message looks reasonable. Since the post can be longer, it might be good to also include a link to https://github.com/autokey/autokey/releases/ or a git clone command. That might be overkill though.

I don't think users would focus on it being "built". Maybe something like Version x.x.x has been released and is available at https://github.com/autokey/autokey/releases/. Installation instructions can be found  at https://github.com/autokey/autokey/wiki/Installing. (It's also on PyPI, but anyone who wants it that way will probably already know that.)

**Comment by Elliria** [comment]
@danballweg, we don't currently have a secret for Gitter in the AutoKey repository's settings.

@josephj11, if you've got access to the AutoKey Gitter account, you'd need to:
1. Log in to Gitter as AutoKey.
2. Generate a Gitter token.
3. Open the AutoKey repository's settings.
4. Create a new repository secret that uses the token.
5. Use that secret in a GitHub action.

Then, any messages that it sends will originate from AutoKey.

If you don't have access to the AutoKey Gitter account, you could do the same thing by logging into Gitter as yourself, instead, and using your Gitter token in the AutoKey secret, and then any messages that it sends would be from you.

**Comment by josephj11** [comment]
@Elliria I'll look into it ASAP. Thatnks for the explicit instructions.

**Comment by josephj11** [comment]
I spent over an hour on that and got nowhere. I'll send you the AI session privately because I did manage to get the Matrix access key for me and I probably shouldn't post that in the open.

**Comment by danballweg** [comment]
I would be happy to jump on a chat if needed.

Here is how you create the secret in the repo:
1. Go to Settings for the repo (along the top next to Insights)
2. `Secrets and variables` (in the list on the left)
3. `Actions` (in the sub menu under Secrets)
4. Click `New Repository Secret` (big green button)
5. For Name put `GITTER_TOKEN`
6. Paste the token you got into the `Secret` section
7. Click `Add secret` to save it

I would prefer to put it under environment secrets and setup an environment so that it can only be used on protected branches, but that would require clicking around in some dangerous portions of the repo. Adding it as just a secret means that it is usable by actions on PRs (which means a malicious user could use it to spam the gitter server if they wanted).  I would be happy to jump on a zoom and guide someone with admin through the process if you want.  otherwise it is usually accepted as a risk, and we can go through that process later if anyone decides to go spamming the gitter channel.

**Comment by Elliria** [comment]
> [@Elliria](https://github.com/Elliria) I'll look into it ASAP. Thatnks for the explicit instructions.

There's no hurry. We can't use it yet anyway since we can't put any PRs through until the branches are in synchronization.

**Comment by Elliria** [comment]
> I spent over an hour on that and got nowhere. I'll send you the AI session privately because I did manage to get the Matrix access key for me and I probably shouldn't post that in the open.

To generate the Gitter access token:
1. Log into Gitter.
2. Click the gear in the bottom left.
3. Click **All settings**.
4. Click **Help & About**.
5. Scroll down to the **Advanced** section.
6. Click **Access Token**.
7. Copy the token.

Then, once we figure out which type of secret we'd like to use, we can move forward with @danballweg's steps.

**Comment by Elliria** [comment]
> I would be happy to jump on a chat if needed.

That's very sweet of you.

> Here is how you create the secret in the repo:

I followed the first four steps and it looks straightforward to me. Now, it's just a matter of figuring out which type of secret we should make.

> I would prefer to put it under environment secrets and setup an environment so that it can only be used on protected branches, but that would require clicking around in some dangerous portions of the repo.

It looks like you're good at writing step-by-step instructions, so that shouldn't be an issue. I took a look in Environments and we've just got one for the AutoKey documentation and would need to create a new one for this.

> Adding it as just a secret means that it is usable by actions on PRs (which means a malicious user could use it to spam the gitter server if they wanted).

Isn't that a non-issue since all PRs have to be approved? It seems that we'd realize they were trying to do that by looking over the code in the PR and could just reject it without ever running it if they did.

> I would be happy to jump on a zoom and guide someone with admin through the process if you want.

Another option for chat would be the AutoKey IRC channel.

**Comment by danballweg** [comment]
I forgot about requiring approval for running workflows.  I am used to all workflows just running, and those few secrets that are a little more dangerous only run on the main branch after merge (mostly all internal repos to the company). With approvals on for running workflows then I wouldn't worry about the environment bit. standard repo secret is just fine

**Comment by Elliria** [comment]
Here are a few landmarks that you'll probably find useful:
* The [build.yml](https://github.com/autokey/autokey/blob/master/.github/workflows/build.yml) workflow is automatically triggered when AutoKey is built (when a release tag is created). 
* The [python-test.yml](https://github.com/autokey/autokey/blob/master/.github/workflows/python-test.yml) workflow is automatically triggered whenever a PR is submitted.
* The [pages.yml](https://github.com/autokey/autokey/blob/master/.github/workflows/pages.yml) workflow is automatically triggered whenever a PR is submitted and nothing happens unless we made changes to [AutoKey's legacy documentation](https://autokey.github.io/autokey/index.html).
* [AutoKey's current documentation](https://autokey.github.io/index.html) is in [its own repository](https://github.com/autokey/autokey.github.io) and has its own [pages.yml](https://github.com/autokey/autokey.github.io/blob/master/.github/workflows/pages.yml) workflow that we run manually when changes have been made to it. It uses Sphinx and fetches the documentation from within the AutoKey code.

I'm assuming that what you two are putting together here will only run when there's a release, so it would be added to the [build.yml](https://github.com/autokey/autokey/blob/master/.github/workflows/build.yml) workflow, right?

**Comment by josephj11** [comment]
I'm just following along here. IF I need to do anything specific, let me know. I tried some tihings on the Gitter end with limited progress. I never found the Access Token on option on Gitter. I ended up inspecting  the web page and pulling it out of Element page (with a bunch of AI help).

**Comment by danballweg** [comment]
I will have some time this weekend to really dig into the branch merging issue.  I will see if i can get v0.96.1 to the point of being able to be merged to master and released.  Then we can test this out at the same time.

> I'm assuming that what you two are putting together here will only run when there's a release, so it would be added to the [build.yml](https://github.com/autokey/autokey/blob/master/.github/workflows/build.yml) workflow, right?

Yes, this will be an addition to the build.yml workflow and run when there is a new release.

**Comment by josephj11** [comment]
0.96.1 is a twinkle in my eyes ATM. The problem (after getting the branches in sync) is that the Wayland stuff is already in develop and that's for 0.97.0.

So, to  create 0.96.1, we would have to clone master and cherry pick things from develop into it. AFAIK, there's no list... of which things that would be, so we'd have to review everything...

Depending on how much work that would be, we have to decide if it's worth the effort or if we shoud just work on 0.97.0. This hasn't really been discussed yet.

This could potentially get master out of sync with develop again, so we'd have to figure out how to deal with that in advance.

**Comment by Elliria** [comment]
It sounds to me like @danballweg may be able to get the branches sorted out so that we can start processing PRs again, in which case, either or both releases could be worked on. Stability has to come first and then all that Wayland code that has been contributed will finally get a chance to become a part of the project. This is a very nice twinkle.

**Comment by josephj11** [comment]
@Elliria Hope you're right! 

The changes that would go into 0.96.1 are probably split between the main branch and develop.

The ones from develop would have to be cherry picked into the new release because all the Wayland changes are aready in develop.

I don't have any idea which PRs... that would be. We'd have to review everything to figure it out. My concern is that anything cherry picked from  develop and put into 0.96.1 might make develop and 0.96.1 out of sync again. I don't really understand  how that would affect git.

**Comment by Elliria** [comment]
Ah, I see what you mean. We got ahead or ourselves. 0.97.0 it is, then. That would be a lot less messy.

**Comment by josephj11** [comment]
That would be quite a reversion! LOL That was before I got here! I assume that was a typo ... 0.97.0.

The problem with 0.97.0 is that it is largely untested. I don't even know if it's feature complete.

I'm pretty sure visgrep() won't work (it has a calling sequence bug now, but nobody uses it, so it wasn't addressed). I'm pretty sure it hasn't been disabled or removed. I have a couple of uses for it,  but never got around to implementing them.

**Comment by Elliria** [comment]
Ack, fixed. That's what I get for posting when bedtime is fast approaching. I suspect there will be plenty of folks willing to do the testing as soon as we've got something for them. Maybe visgrep works with [autokey-shell](https://github.com/autokey/autokey/blob/4b87b32452cd41d2b98341db220e99cf4f9e34d6/new_features.rst#running-autokey-scripts-interactively-on-a-shell). I haven't tried it.

**Comment by josephj11** [comment]
@Elliria  Visgrep will work just fine, iff you call it externally using subprocess... The problems are our API calling sequence. But, most importantly, it's totally dependent on x11 and the Xautomation project has been abandoned for years, so, if we want it, we have to fork it as our own branch and convert it to Wayland.

I'd like nothing better, but it seems a tad ambitious. LOL

If we found somebody who knew how, I'd bet we could throw AutoKey and Xautomation and x11 into a flatpak and have it all work iff it could reach out of the sandbox to the rest of the system. xlibre is being developed, so we could port to that for a current x11 environment. Not quite rocket science, but pretty close.

**Comment by Elliria** [comment]
If it's dependent on X, then it may as well be left to collect dust. Wayland is taking over, so it would probably be better to look for a Python (**PyAutoGUI**, maybe) or Wayland-specific solution for something like that.

From Google AI:
> For scripting screenshots and automating screen interactions in Wayland, use **grim** for capture, **slurp** for region selection, **wl-clipboard** for clipboard integration, and potentially **swappy** or **satty** for editing, but for interacting with on-screen elements like buttons (which involves simulating input), you'll need tools like **ydotool**, **wtype**, or Python libraries with dbus interaction (XDG Portals), as Wayland's security restricts direct window control.

**Comment by josephj11** [comment]
Those sound useful, but what Xauomation does is actually find images on the screen and return their coordinates so you can use them, e.g. click on them. It even lets you find images within images when context is important, such as looking for an Ok button on the screen when more than one might be present. (The AutoKey API  wasn't coded for that option at all.)

My script to automate the chrome print dialog currently uses a couple of hard-coded coordinates to click on buttons when it doesn't know for sure where the cursor is to start with. I tried writing it without that and failed a number of times because of multiple edge cases. Also, if the browser isn't maximized, everything is somewhere else and my script breaks. There are a couple of other print dialogs where I can tab all day and never get to the button I need to press.

It's probably a lost cause, but it's a feature I really want.

**Comment by josephj11** [comment]
Actually, I did a little preliminary research on replacing visgrep() and it doesn't look anywhere near as hard as I thought it might be. It still requires reasonable Python skills and a bunch of effort, but there don't seem to be any obvious major roadblocks other than motivating someone to do it. It seems way too OT to post it here.

**Comment by danballweg** [comment]
Correct me if I am wrong, but in order to release we have to merge things to master.  If we were to be able to get all the changes for 0.96.1 cherry-picked, then we would have master branch at 0.96.1, and develop would be just the changes that didn't get released in 0.96.1

Once the cherry-picked commits are merged into master, then the diff between develop and master would shrink.

I am all for just jumping to 0.97.0 if that stuff is close though.  I could take a look at what is left for that if we would rather focus efforts on getting that released.  Do we have a burn-down list for things that need to be completed in order to release the Wayland stuff?

I got setup over the weekend to be able to run tests, and play around a bit.  I found the changelog, and it seems like a number of things are in master that are marked for the next release [https://github.com/autokey/autokey/blob/87ace10981b00bee1e5069fb7b1b5da9770a80bb/CHANGELOG.rst#L5](CHANGELOG.rst)
We could tack on a couple little bug fixes and run that out as 0.96.1 if you wanted.

**Comment by josephj11** [comment]
My command of git/GitHub is  less than minimal, but what I think is going on is:

There are some changes that were made directly to master and not to develop. For  those, we have to make sure that they get applied to develop - without blowing away anything that was changed in develop that we need for 0.97.0.

There are some (as yet unidentified) changes to develop that should be in 0.96.1 which we need to cherry pick into master for 0.96.1. Hopefully, when we do that, that will reduce the "diff" between master and develop as you said.

Once the changes above are cherry picked into master (and my pet issue, #600 is implemented), theoretically, 0.96.1 should be ready to build and release. At that point we would need to review all the outstanding differences  and make sure that whatever we're doing with 0.96.1 isn't going to increase our technical debt (at least significantly). We would make a beta release and ask the community to test it. When it's ready, we woulld release it and notify the packagers.

0.97.0
Sam, and to some extent, bluedrink9, developed the Wayland code and a bunch of code refactoring and possibly other improvents/bug fixes. I am not aware  of any list of exactly what was done and what remains to be done although there are a few Discussion issues that  were opened with a lot of helpful information in them.

We need to review it all and see if it's feature complete (without visgrep(), which we know hasn't been addressed) and figure out what development, if any, is needed to make it a viable release. Since it's unlikely (in the short term, if ever) that anyone is going to rewrite Xautomation for Wayland, we will need to disable it in our code and documentation. I'd hate to see it completely removed, but that might be a reasonabe thing to do.

Once that's done, we have to make a beta release, make some sort of testing criteria, and test it internally until it looks mostly ready. Then, we can ask the community to start testing it. That and any fixes  that are subsequenty made will give us a    release candidate that we can announce a bit more broadly. Once we're confident that it's solid enough, we release 0.97.0 and send out notices to various packagers that it's available.

**Comment by Elliria** [comment]
Some of that is right, but not all. [This post](https://github.com/autokey/autokey/discussions/823#discussioncomment-11285427) provides the steps that are designed to take us from **development** to **beta** to **release**.

We didn't follow them properly (I include myself as one of the guilty parties in that), so we've got a mess on our hands that needs to be sorted out. That has to come first. Our initial goal should simply be to achieve code stability so that pull requests can be put through.

At that point, we'll be able to test out some of the fixes and updates in the pull requests and also the Wayland code that's been waiting to be merged in.

I'd say it's pretty ambitious for us to be thinking of aiming for version **0.97.0**, but in the event that you'd like to, [this is its milestone](https://github.com/autokey/autokey/issues?q=is%3Aissue%20state%3Aopen%20milestone%3A%22Feature%20release%200.97.0%22).

The **CHANGELOG.rst** file is another problem. Here's the one on [master](https://github.com/autokey/autokey/blob/master/CHANGELOG.rst) and here's the one on [develop](https://github.com/autokey/autokey/blob/develop/CHANGELOG.rst). The differences in those will need to be sorted out. Each contains changes for the next version that were made on the branch that they're on. Part of cleaning this mess up will be comparing them and cherry-picking the changes that we need or want to keep from **master** to **develop**.

We should probably move this to a [discussion](https://github.com/autokey/autokey/discussions). Perhaps [this one](https://github.com/autokey/autokey/discussions/823) would do, since it's already talking about this stuff.
